pulp_container/app/registry_api.py

```diff
@@ -564,18 +564,95 @@ def create(self, request, path):
         """
         Create a new upload.
 
-        Note: We do not support monolithic upload.
         """
         _, repository = self.get_dr_push(request, path, create=True)
 
         if self.tries_to_mount_blob(request):
             response = self.mount_blob(request, path, repository)
+            return response
+        elif digest := request.query_params.get("digest"):
+            # if the digest parameter is present, the request body will be
+            # used to complete the upload in a single request.
+            self.single_request_upload(request, path, repository, digest)
         else:
             upload = models.Upload(repository=repository, size=0)
             upload.save()
             response = UploadResponse(upload=upload, path=path, request=request)
+            return response
 
-        return response
+    def create_single_chunk_artifact(self, chunk):
+        with transaction.atomic():
+            # 1 chunk, create artifact right away
+            with NamedTemporaryFile("ab") as temp_file:
+                temp_file.write(chunk.read())
+                temp_file.flush()
+
+                uploaded_file = PulpTemporaryUploadedFile.from_file(
+                    File(open(temp_file.name, "rb"))
+                )
+            try:
+                artifact = Artifact.init_and_validate(uploaded_file)
+                artifact.save()
+            except IntegrityError:
+                artifact = Artifact.objects.get(sha256=artifact.sha256)
+                artifact.touch()
+            return artifact
+
+    def create_blob(self, artifact, digest):
+        with transaction.atomic():
+            try:
+                blob = models.Blob(digest=digest)
+                blob.save()
+            except IntegrityError:
+                blob = models.Blob.objects.get(digest=digest)
+                blob.touch()
+            try:
+                blob_artifact = ContentArtifact(
+                    artifact=artifact, content=blob, relative_path=digest
+                )
+                blob_artifact.save()
```

Comment:
I think, this should be part of the first try. The blob cannot exist without it's content artifact. So finding the blob in 608 means the ContentArtifact must already be in place.

Reply:
you're right, i will change this.
I mostly moved around the code that was repeated in few place without looking too thoroughly

Reply:
Actually, I will not. Here is the reason to it! https://github.com/pulp/pulp_container/commit/c06c025d2193ecec870edaf0f1fd6fde266aac3e
Should I rather leave a comment in the code?